### ISSUE TITLE
gRPC is incompatible with tls.Listener

### ISSUE THREAD

**Issue by realh** [grpc/grpc-go#7313]
### What version of gRPC are you using?
1.64.0

### What version of Go are you using (`go version`)?
1.21.1

### What operating system (Linux, Windows, …) and version?
Ubuntu 22.04 arm64

### What did you do?
I want to serve gRPC and other services on the same address and port, using cmux, but gRPC does not work with a `Listener` created by `tls.Listen()`, whether I pass that Listener directly to `grpc.Serve()` or use it to create a `CMux` and pass a matching Listener to grpc. If I use `net.Listen()` instead, and add TLS credentials to grpc's `ServerOption`s, it works normally, whether the Listner is connected directly to grpc or via a cmux matcher. Unfortunately, none of cmux's matchers except `Any()` work with undecrypted TLS, rendering it useless.

Edit: I also tested the `Listener` created by `tls.Listen()` with the HTTP1 service I want to run at the same time, which is implemented with an `http.Server`. With this server the Listener works correctly, whether it's used directly or via a cmux matcher. It seems to be a specific incompatibility between crypto/tls and grpc.

### What did you expect to see?
I thought that grpc should work with a `Listener` created by the `tls` module, letting `tls` handle the en/decryption. Or at least report an error if it didn't.

### What did you see instead?
Nothing server-side. No errors are reported, no handler is called, and the client fails with an "UNAVAILABLE" error.

More details are available in [a conversation I had with ChatGPT](https://chatgpt.com/share/c21e1e21-6e28-4072-8db7-d47f578ee353). I know it's not considered reliable, but I think it did a pretty good job here, considering the information it had available.

**Comment by purnesh42H** [comment]
@realh I will take a look at this

**Comment by realh** [comment]
Actually, I think it was a problem with my code after all. I ran into a similar problem after adopting a different approach, simply running gRPC on a different port from the other service. This time I eventually realised it was because I was loading `cert.pem` instead of `fullchain.pem`. That works when the client is a browser, because browsers cache well-known certificates, but gRPC clients don't. Thanks and apologies for your time.
